### PR TITLE
[feat] L2 fee calculator

### PR DESCRIPTION
**Pull request by ytham**
- L2 fee calculator:
  - OpStack supported
  - Arbitrum WIP
  - Scroll WIP
- Base fee calculator tests

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -9,9 +9,11 @@ on:
   pull_request:
 
 env:
+  PRIVATE_KEY_ANVIL: ${{ secrets.PRIVATE_KEY_ANVIL }}
   PRIVATE_KEY_SEPOLIA: ${{ secrets.PRIVATE_KEY_SEPOLIA }}
   PROVIDER_URI_SEPOLIA: ${{ secrets.PROVIDER_URI_SEPOLIA }}
   PROVIDER_URI_GOERLI: ${{ secrets.PROVIDER_URI_GOERLI }}
+  PROVIDER_URI_84532: ${{ secrets.PROVIDER_URI_84532 }}
```
Comment (yi-sun) [inline .github/workflows/test.yml:16]: Why use the chainID instead of name here?

File: client/src/lib/paymentCalc.ts [hunk 2]
```diff
@@ -1,58 +1,108 @@
...
+    // on OP Stack
+    // projectedCallbackCost = 
+    //   maxFeePerGas * (callbackGasLimit + proofVerificationGas) +    
+    //   AXIOM_PROOF_CALLDATA_BYTES * 16 * (L1BlockAttributes.baseFeeScalar * L1BlockAttributes.basefee + 
+    //   L1BlockAttributes.blobBaseFeeScalar / 16 * L1BlockAttributes.blobBaseFee) / 1e6
+    const baseFeeScalar = await getOpStackL1AttributesValue(publicClient, "baseFeeScalar", []);
+    const baseFee = await getOpStackL1AttributesValue(publicClient, "basefee", []);
+    const blobBaseFeeScalar = await getOpStackL1AttributesValue(publicClient, "blobBaseFeeScalar", []);
+    const blobBaseFee = await getOpStackL1AttributesValue(publicClient, "blobBaseFee", []);
+    return maxFeePerGas * (callbackGasLimit + proofVerificationGas) +
```
Comment (yi-sun) [inline client/src/lib/paymentCalc.ts:86]: I didn't know about this before, but found: https://viem.sh/op-stack/actions/estimateL1Gas

Comment (yi-sun) [reply client/src/lib/paymentCalc.ts:86]: I think this will be more robust than doing the computation directly

**Review by yi-sun — CHANGES_REQUESTED** [review]
Probably better to use https://viem.sh/op-stack/actions/estimateL1Gas

File: .github/workflows/test.yml (hunk 1 above)
Comment (ytham) [reply .github/workflows/test.yml:16]: this is part of the beginning of a broader change across the SDK repos where we can more easily keep provider URIs in line w/ chain IDs and reduce confusion/errors as we onboard more chains.

Comment (ytham) [reply .github/workflows/test.yml:16]: note that this only affects our tests/CI

**Comment by ytham** [comment]
@yi-sun ready for re-review

File: .github/workflows/test.yml (hunk 1 above)
Comment (yi-sun) [reply .github/workflows/test.yml:16]: Sounds good

File: client/src/lib/paymentCalc.ts [hunk 3]
```diff
@@ -1,58 +1,104 @@
...
+  } else {
+    // overrideAxiomQueryFee = AXIOM_QUERY_FEE + projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas)
+    overrideAxiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI + 
+      projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas);
   }
-  axiomQueryFee = BigInt(axiomQueryFee ?? 0);
-  if (axiomQueryFee === 0n) {
-    axiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+
+  // Get axiomQueryFee from contract if on ethereum, otherwise use L2 equation
```
Comment (yi-sun) [inline client/src/lib/paymentCalc.ts:44]: Can we re-arrange the logic here to more explicitly branch on the chainId?  It's hard to follow the flow in the L1 and L2 case right now because they are mixed

**Review by yi-sun — CHANGES_REQUESTED** [review]

File: client/src/lib/paymentCalc.ts [hunk 4]
```diff
@@ -1,58 +1,104 @@
...
-  axiomV2QueryAddr: string,
+  chainId: string,
   publicClient: PublicClient,
-  options?: AxiomV2QueryOptions
+  options?: AxiomV2ClientOptions,
 ): Promise<bigint> {
+  const axiomV2QueryAddr = options?.overrides?.queryAddress ?? getAxiomV2QueryAddress(chainId);
+
+  // Convert callback gas limit to wei
+  const callbackGasLimit = BigInt(options?.callbackGasLimit ?? ClientConstants.DEFAULT_CALLBACK_GAS_LIMIT);
```
[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:17]: depending on how the calculatePayment function is called, the `options` argument can be undefined. when `calculatePayment` called inside the client SDK, options is filled out the value passed into it, but if it's called externally, the caller may not have `options` set

[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:17]: and yes, not doing any conversion just handling when values are undefined

[the remark this answers is not in the document]
Comment (yi-sun) [reply client/src/lib/paymentCalc.ts:18]: When would it be called externally?

[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:18]: currently in tests, but it's exported by the client package for devs to call if they want to pass in their own values for their own logic. we can definitely make options non-optional and have devs to pass in values instead of using defaults if nothing is passed in

**Comment by ytham** [comment]
@yi-sun ready for review again

File: client/src/lib/paymentCalc.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (yi-sun) [reply client/src/lib/paymentCalc.ts:18]: Yeah, I think that's safer (and we can have default values exported as constants).

File: client/src/lib/paymentCalc.ts [hunk 5]
```diff
@@ -1,56 +1,115 @@
...
+
+    // max(overrideAxiomQueryFee, axiomQueryFee)
+    if (overrideAxiomQueryFee > axiomQueryFee) {
+      axiomQueryFee = overrideAxiomQueryFee;
+    }
+
+    // Calculate payment
+    let payment = axiomQueryFee + maxFeePerGas * (proofVerificationGas + callbackGasLimit);
+    const minimumPayment = projectedCallbackCost + ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+    if (payment < minimumPayment) {
```
Comment (yi-sun) [inline client/src/lib/paymentCalc.ts:65]: I think we'll want to error in this case, since it means something was misconfigured

**Review by yi-sun — CHANGES_REQUESTED** [review]

File: client/src/lib/paymentCalc.ts [hunk 6]
```diff
@@ -1,56 +1,115 @@
...
+  callbackGasLimit: bigint,
+  proofVerificationGas: bigint,
+): Promise<bigint> {
+  if (isOpStackChain(chainId)) {
+    // on OP Stack
+    // projectedCallbackCost = 
+    //   maxFeePerGas * (callbackGasLimit + proofVerificationGas) +    
+    //   AXIOM_PROOF_CALLDATA_LEN * 16 * (L1BlockAttributes.baseFeeScalar * L1BlockAttributes.basefee + 
+    //   L1BlockAttributes.blobBaseFeeScalar / 16 * L1BlockAttributes.blobBaseFee) / 1e6
+    const l1Fee = await readContractValue(
```
[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:90]: yeah, it looks like estimateL1Fee calls `getL1Fee` under the hood: https://github.com/wevm/viem/blob/98f46303662ccc7f69da8f0c35dc7f9e5c52cd56/src/chains/opStack/actions/estimateL1Fee.ts#L116

File: client/src/lib/paymentCalc.ts [hunk 7]
```diff
@@ -1,56 +1,115 @@
...
+    }
+
+    // Calculate payment
+    let payment = axiomQueryFee + maxFeePerGas * (proofVerificationGas + callbackGasLimit);
+    const minimumPayment = projectedCallbackCost + ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+    if (payment < minimumPayment) {
+      payment = minimumPayment;
+    }
+    return payment;
+  } else {
```
[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:69]: ah, so this calculatePayment function is still used for all chains but we explicitly branch here: https://github.com/axiom-crypto/axiom-sdk-client/pull/78/files/3cb06ba859f2584d82e1fbc548fbccf24aa3b919#diff-5bca24442066d1d1b8d5ca80558df169f1725d57c1d06ab94f0e870c691a9140R43

[the remark this answers is not in the document]
Comment (yi-sun) [reply client/src/lib/paymentCalc.ts:69]: I mean in the branch below, if a chain is not OP Stack, Arbitrum, Scroll, or mainnet, we should error

[the remark this answers is not in the document]
Comment (ytham) [reply client/src/lib/paymentCalc.ts:69]: kk, sounds good

**Comment by ytham** [comment]
@yi-sun should be good to go now; lmk there are any other changes i should make

File: client/src/lib/paymentCalc.ts [hunk 8]
```diff
@@ -1,56 +1,117 @@
...
-  options?: AxiomV2QueryOptions
+  options: AxiomV2ClientOptions,
 ): Promise<bigint> {
+  const axiomV2QueryAddr = options.overrides?.queryAddress ?? getAxiomV2QueryAddress(chainId);
+
+  // Get callback gas limit
+  const callbackGasLimit = BigInt(options.callbackGasLimit ?? ClientConstants.DEFAULT_CALLBACK_GAS_LIMIT);
+
+  // Get maxFeePerGas
+  const maxFeePerGas = BigInt(options.maxFeePerGas ?? ClientConstants.DEFAULT_MAX_FEE_PER_GAS_WEI);
```
Comment (yi-sun) [inline client/src/lib/paymentCalc.ts:21]: Just realized -- we'll want to make the ClientConstants depend on `chainId`, e.g. this `DEFAULT_MAX_FEE_PER_GAS_WEI` should be quite different on Base / mainnet

File: client/src/lib/paymentCalc.ts [hunk 9]
```diff
@@ -1,56 +1,117 @@
...
-  if (proofVerificationGas === 0n) {
-    proofVerificationGas = ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS;
-  }
+  const proofVerificationGas = await readContractValue(
+    publicClient,
+    axiomV2QueryAddr,
+    getAxiomV2Abi(AbiType.Query),
+    "proofVerificationGas",
+    [],
+    ClientConstants.FALLBACK_PROOF_VERIFICATION_GAS
```
Comment (yi-sun) [inline client/src/lib/paymentCalc.ts:30]: Similar per-chainId constants

File: client/src/types.ts [hunk 10]
```diff
@@ -19,7 +19,6 @@ export interface AxiomV2ClientConfig<T> {
   provider: string;
   callback: AxiomV2CallbackInput;
   privateKey?: string;
-  version?: string;
```
Comment (yi-sun) [inline client/src/types.ts:22]: Why was this removed?  Same for the `privateKey` field below

**Review by yi-sun — CHANGES_REQUESTED** [review]

File: client/src/types.ts (hunk 10 above)
Comment (ytham) [reply client/src/types.ts:22]: they're not used and are also not in the docs; i can make a separate PR to remove if preferred

Comment (yi-sun) [reply client/src/types.ts:22]: OK, this is fine, but in general it makes the PR pretty hard to review / we should explicitly separate cleanup / refactoring from feature additions.

File: client/src/axiom/axiom.ts [hunk 11]
```diff
@@ -163,7 +163,7 @@ export class Axiom<T> {
     }
     const clientOptions: AxiomV2ClientOptions = {
       ...this.options,
-      callbackGasLimit: this.options?.callbackGasLimit ?? ClientConstants.CALLBACK_GAS_LIMIT,
+      callbackGasLimit: this.options?.callbackGasLimit ?? Number(getChainDefaults(this.config.chainId).callbackGasLimit),
```
Comment (yi-sun) [inline client/src/axiom/axiom.ts:166]: Why do we need the `Number` here?

File: client/src/constants.ts [hunk 12]
```diff
@@ -1,9 +1,21 @@
+import { ChainDefaults } from "./types";
+
 export const ClientConstants = Object.freeze({
   MIN_MAX_FEE_PER_GAS: 5000000000n,
```
Comment (yi-sun) [inline client/src/constants.ts:4]: We will need to make this chain-dependent as well, since this is too high for OP Stack chains.

File: client/src/constants.ts [hunk 13]
```diff
@@ -1,9 +1,21 @@
...
+});
 
-  FALLBACK_PROOF_VERIFICATION_GAS: 420000n,
-  FALLBACK_AXIOM_QUERY_FEE_WEI: 3000000000000000n,
-  DEFAULT_CALLBACK_GAS_LIMIT: 100000,
-  DEFAULT_MAX_FEE_PER_GAS_WEI: "25000000000",
-})
\ No newline at end of file
+export const BaseDefaults: Readonly<ChainDefaults> = Object.freeze({
+  maxFeePerGasWei: 25000000000n,
```
Comment (yi-sun) [inline client/src/constants.ts:17]: We can use these default values for Base:
```
  MIN_MAX_FEE_PER_GAS: 10000000n,
  CALLBACK_GAS_LIMIT: 100000,
  FALLBACK_PROOF_VERIFICATION_GAS: 420000n,
  FALLBACK_AXIOM_QUERY_FEE_WEI: 3000000000000000n,
  DEFAULT_MAX_FEE_PER_GAS_WEI: "10000000",
```

**Review by yi-sun — CHANGES_REQUESTED** [review]

**Comment by ytham** [comment]
@yi-sun ready for review again

File: client/src/axiom/utils.ts [hunk 14]
```diff
@@ -95,15 +86,16 @@ export async function getMaxFeePerGas(axiom: AxiomSdkCore, overrides?: AxiomV2Cl
       args: [],
     }) as bigint;
     if (contractMinMaxFeePerGas === 0n) {
-      contractMinMaxFeePerGas = ClientConstants.MIN_MAX_FEE_PER_GAS;
+      contractMinMaxFeePerGas = getChainDefaults(chainId).maxFeePerGasWei;
```
Comment (yi-sun) [inline client/src/axiom/utils.ts:89]: should be `minMaxFeePerGasWei`

File: client/src/axiom/utils.ts [hunk 15]
```diff
@@ -95,15 +86,16 @@ export async function getMaxFeePerGas(axiom: AxiomSdkCore, overrides?: AxiomV2Cl
...
     }
     if (providerFeeData > contractMinMaxFeePerGas) {
       return providerFeeData.toString();
     }
     console.log(`Network gas price below threshold. Using contract-defined minimum maxFeePerGas of ${contractMinMaxFeePerGas.toString()}`);
```
Comment (yi-sun) [inline client/src/axiom/utils.ts:94]: the log text is inconsistent, I think it should be "minimum minMaxFeePerGas"

File: client/src/axiom/utils.ts [hunk 16]
```diff
@@ -95,15 +86,16 @@ export async function getMaxFeePerGas(axiom: AxiomSdkCore, overrides?: AxiomV2Cl
...
     return contractMinMaxFeePerGas.toString();
   } catch (e) {
-    console.log(`Unable to read minMaxFeePerGas from contract, returning default value of ${ClientConstants.DEFAULT_MAX_FEE_PER_GAS_WEI}`);
-    return ClientConstants.DEFAULT_MAX_FEE_PER_GAS_WEI.toString();
+    const defaultMaxFeePerGas = getChainDefaults(chainId).maxFeePerGasWei.toString();
+    console.log(`Unable to read minMaxFeePerGas from contract, returning default value of ${defaultMaxFeePerGas}`);
```
Comment (yi-sun) [inline client/src/axiom/utils.ts:98]: Similar to above, this should be `minMaxFeePerGas` instead of `maxFeePerGas`

**Review by yi-sun — CHANGES_REQUESTED** [review]

**Comment by ytham** [comment]
@yi-sun comments addressed

**Review by yi-sun — APPROVED** [review]
